frontend/packages/sdk/src/contracts/PinkContract.ts

```diff
@@ -0,0 +1,287 @@
+    const seed = hexToU8a(hexAddPrefix(randomHex(32)));
+    const pair = sr25519KeypairFromSeed(seed);
+    const [sk, pk] = [pair.slice(0, 64), pair.slice(64)];
+
+    const queryAgreementKey = sr25519Agree(
+      hexToU8a(hexAddPrefix(this.phatRegistry.remotePubkey)),
+      sk
+    );
+
+    const asyncMethod = async (origin: string | AccountId | Uint8Array): Promise<ContractCallOutcome> => {
+      // @ts-ignore
+      const cert = await signCertificate({ pair: origin, api });
+      const payload = api.createType("InkQuery", {
+        head: {
+          nonce: hexAddPrefix(randomHex(32)),
+          id: this.address,
+        },
+        data: {
+          InkMessage: message.toU8a(params),
+        },
+      });
+      const data = await pinkQuery(api, this.phatRegistry.phactory, pk, queryAgreementKey, payload.toHex(), cert);
+      const { debugMessage, gasConsumed, gasRequired, result, storageDeposit } = api.createType<ContractExecResult>(
+        "ContractExecResult",
+        (
+          api.createType("InkResponse", hexAddPrefix(data)).toJSON() as {
+            result: { ok: { inkMessageReturn: string } };
+          }
+        ).result.ok.inkMessageReturn
+      );
+      return {
+        debugMessage: debugMessage,
+        gasConsumed: gasConsumed,
+        gasRequired: gasRequired && !convertWeight(gasRequired).v1Weight.isZero() ? gasRequired : gasConsumed,
+        output: result.isOk && message.returnType
+          ? this.abi.registry.createTypeUnsafe(message.returnType.lookupName || message.returnType.type, [result.asOk.data.toU8a(true)], { isPedantic: true })
+          : null,
+        result,
+        storageDeposit
+      }
+    }
+
+    return {
+      send: this._decorateMethod((origin: string | AccountId | Uint8Array) => from(asyncMethod(origin)))
+    };
+  };
+
+  #inkCommand = (messageOrId: AbiMessage | string | number, { gasLimit = BN_ZERO, storageDepositLimit = null, value = BN_ZERO }: ContractOptions, params: unknown[]): SubmittableExtrinsic<'promise'> => {
+    const api = this.api as ApiPromise
+
+    // Generate a keypair for encryption
+    // NOTE: each instance only has a pre-generated pair now, it maybe better to generate a new keypair every time encrypting
+    const seed = hexToU8a(hexAddPrefix(randomHex(32)));
+    const pair = sr25519KeypairFromSeed(seed);
+    const [sk, pk] = [pair.slice(0, 64), pair.slice(64)];
+
+    const commandAgreementKey = sr25519Agree(hexToU8a(this.contractKey), sk);
+
+    const asyncMethod = (dest: AccountId, value: BN, gas: { refTime: BN }, storageDepositLimit: BN | undefined, encParams: Uint8Array) => {
+      // @ts-ignore
+      const payload = api.createType("InkCommand", {
+        InkMessage: {
+          nonce: hexAddPrefix(randomHex(32)),
+          // FIXME: unexpected u8a prefix
+          message: api.createType("Vec<u8>", encParams).toHex(),
```

Comment:
It's suggested to pass a hex string when constructing `Vec<u8>` fields in polkadot.js. So here I'd suggest to call `u8aToHex()` from polkadotjs utils lib.

Reply:
I'm revert this change because I found the encoded messages decode failed in PRuntime.

Reply:
Which change? Btw, `api.createType("Vec<u8>", hexData).toHex() == hexData` is always true. It's suggested to use `u8aToHex(encParams)` here.

Reply:
As I checked, `api.createType("Vec<u8>", hexData).toHex() != u8aToHex(encParams)`. I'm tried print these two into console:

```javascript
console.log('createType', api.createType("Vec<u8>", encParams).toHex())
console.log('u8aToHex', u8aToHex(encParams))
```

And got:
```
createType 0x0df6030d48426164676531363833333830393133333932
u8aToHex 0x5c0df6030d48426164676531363833333830393133333932
```

I guess for `Vec`, it will insert prefix `5c` for the encoded result.

Reply:
`5c` is the scale prefix, indicating the length of the array. Apparently polkadotjs has a weird API for `api.createType("Vec<u8>", data)`

- If `data` is u8a, apply the scale decode (considering the length prefix)
- If `data` is a hex string, create a `Vec<u8>` with its content
- If `data` is not a hex string, create a `Vec<u8>` with its utf8 encoding

To convert `Vec<u8>` to a hex, there are three ways:

- `data.toHex()`: return the content of the array
- `u8aToHex(data)`: same as before
- `u8aToHex(data.toU8a())`: return the scale encoded bytes array

Reply:
```javascript
let data = api.createType('Vec<u8>', '0xAABB');
console.log({
    toHex: data.toHex(),
    toU8a: u8aToHex(data.toU8a()),
    u8aToHex: u8aToHex(data),
});
// { toHex: '0xaabb', toU8a: '0x08aabb', u8aToHex: '0xaabb' }

const encodedHex = '0x08aabb';
data = api.createType('Vec<u8>', hexToU8a(encodedHex));
console.log({
    toHex: data.toHex(),
    toU8a: u8aToHex(data.toU8a()),
    u8aToHex: u8aToHex(data),
});
// { toHex: '0xaabb', toU8a: '0x08aabb', u8aToHex: '0xaabb' }
```